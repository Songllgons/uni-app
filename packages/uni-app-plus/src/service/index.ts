import * as uni from './api'
import { UniServiceJSBridge } from './bridge'
import { registerApp as __registerApp, getApp } from './framework/app'
import {
  definePage as __definePage,
  registerPage as __registerPage,
  getCurrentPages,
} from './framework/page'

export default {
  uni,
  getApp,
  getCurrentPages,
  __definePage,
  __registerApp,
  __registerPage,
  UniServiceJSBridge,
}
