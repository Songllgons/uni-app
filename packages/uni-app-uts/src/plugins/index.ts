import { init as initAndroidPlugins } from './android'
import { init as initIOSPlugins } from './ios'

export function initAndroid() {
  import('./errorReporting')
  return initAndroidPlugins()
}

export function initIOS() {
  import('./errorReporting')
  return initIOSPlugins()
}
