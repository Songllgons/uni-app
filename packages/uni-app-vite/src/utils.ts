import path from 'path'
export const templateDir = path.resolve(__dirname, '../lib/template/')
export function nvueOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.nvue')
}
