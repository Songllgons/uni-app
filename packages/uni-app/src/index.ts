export * from './ssr'
export * from './api'
export * from './hbx'
export * from './utils'
export * from './apiLifecycle'
export * from './vue'
