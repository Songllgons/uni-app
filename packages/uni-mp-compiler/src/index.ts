import { extend } from '@vue/shared'
import {
  type CodegenResult,
  type ParserOptions,
  type RootNode,
  baseParse,
} from '@vue/compiler-core'

import { baseCompile } from './compile'
import { parserOptions } from './parserOptions'
import type { CompilerOptions } from './options'
export type { CompilerOptions } from './options'
export { findProp } from '@vue/compiler-core'
export type {
  DirectiveNode,
  NodeTransform,
  DirectiveTransform,
  TransformContext,
  SimpleExpressionNode,
} from '@vue/compiler-core'

export { genExpr } from './codegen'
export { rewriteExpression } from './transforms/utils'
export { isForElementNode } from './transforms/vFor'
export { transformOn } from './transforms/vOn'
export { transformModel } from './transforms/vModel'
export * from './runtimeHelpers'

export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

export function compile(
  template: string,
  options: CompilerOptions = {}
): CodegenResult {
  return baseCompile(
    template,
    extend({}, parserOptions, options, {
      directiveTransforms: extend({}, options.directiveTransforms || {}),
    })
  )
}
