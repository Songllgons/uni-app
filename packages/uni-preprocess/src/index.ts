export { preprocess } from './preprocess'
export type {
  PreprocessOptions,
  ProcessContext,
  SourceMapOptions,
} from './preprocess'
