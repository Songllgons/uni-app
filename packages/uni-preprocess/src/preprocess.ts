import MagicString, { type SourceMapOptions } from 'magic-string'
import { matchRecursive } from 'xregexp'

export type { SourceMapOptions } from 'magic-string'

export type ProcessContext = Record<string, any>
export interface PreprocessOptions {
  type?: 'js' | 'html'
  context: ProcessContext
  sourceMap?: SourceMapOptions
}

export function preprocess(source: string, options: PreprocessOptions) {
  let isInPreprocessor = (_offset: number) => false
  if (!source.includes('#endif')) {
    return {
      code: source,
      map: options.sourceMap
        ? new MagicString(source).generateMap(options.sourceMap)
        : null,
      isInPreprocessor,
    }
  }

  const context = options.context || {}
  const s = new MagicString(source)

  // [[start,end]]
  const ranges: [number, number][] = []

  isInPreprocessor = (offset: number) => {
    for (let i = 0; i < ranges.length; i++) {
      const range = ranges[i]
      if (range[0] <= offset && offset < range[1]) {
        return true
      }
    }
    return false
  }

  function preprocessByType(type: PreprocessType) {
    replaceRecursive(
      type,
      source,
      s,
      (s, startMatches, endMatches, include, recurse) => {
        // I need to recurse first, so I don't catch "inner" else-directives
        recurse(include.value, include.start)

        const variant = startMatches.value[1]
        const test = (startMatches.value[2] || '').trim()
        switch (variant) {
          case 'ifdef':
            if (testPasses(test, context)) {
              s.remove(startMatches.start, startMatches.end)
              s.remove(endMatches.start, endMatches.end)
            } else {
              s.remove(startMatches.start, endMatches.end)
            }
            ranges.push([startMatches.start, endMatches.end])
            return
          case 'ifndef':
            if (!testPasses(test, context)) {
              s.remove(startMatches.start, startMatches.end)
              s.remove(endMatches.start, endMatches.end)
            } else {
              s.remove(startMatches.start, endMatches.end)
            }
            ranges.push([startMatches.start, endMatches.end])
            return
          default:
            throw new Error('Unknown if variant ' + variant + '.')
        }
      }
    )
  }
  const type = options.type || 'auto'
  if (type === 'auto' || type === 'js') {
    preprocessByType(TYPES.js)
  }
  if (type === 'auto' || type === 'html') {
    preprocessByType(TYPES.html)
  }
  return {
    code: s.toString(),
    map: options.sourceMap ? s.generateMap(options.sourceMap) : null,
    isInPreprocessor,
  }
}

type Processor = (
  s: MagicString,
  startMatches: MatchGroupLeft,
  endMatches: MatchGroupRight,
  include: MatchGroupMatch,
  recurse: (content: string, startOffset: number) => void
) => void

const startPattern =
  '[ \t]*(?://|/\\*)[ \t]*#(ifndef|ifdef)[ \t]+([^\n*]*)(?:\\*(?:\\*|/))?(?:[ \t]*\n+)?'
const endPattern =
  '[ \t]*(?://|/\\*)[ \t]*#endif[ \t]*(?:\\*(?:\\*|/))?(?:[ \t]*\n)?'

const startPatternHTML =
  '[ \t]*<!--[ \t]*#(ifndef|ifdef|if)[ \t]+(.*?)[ \t]*(?:-->|!>)(?:[ \t]*\n+)?'
const endPatternHTML =
  '[ \t]*<!(?:--)?[ \t]*#endif[ \t]*(?:-->|!>)(?:[ \t]*\n)?'

interface PreprocessType {
  start: {
    pattern: string
    regex: RegExp
  }
  end: {
    pattern: string
    regex: RegExp
  }
}

const TYPES: Record<string, PreprocessType> = {
  js: {
    start: {
      pattern: startPattern,
      regex: new RegExp(startPattern, 'mi'),
    },
    end: {
      pattern: endPattern,
      regex: new RegExp(endPattern, 'mi'),
    },
  },
  html: {
    start: {
      pattern: startPatternHTML,
      regex: new RegExp(startPatternHTML, 'mi'),
    },
    end: {
      pattern: endPatternHTML,
      regex: new RegExp(endPatternHTML, 'mi'),
    },
  },
}

interface MatchGroupItem {
  start: number
  end: number
}
interface MatchGroupMatch extends MatchGroupItem {
  value: string
}
interface MatchGroupLeft extends MatchGroupItem {
  value: RegExpExecArray
}

type MatchGroupRight = MatchGroupLeft

function replaceRecursive(
  type: PreprocessType,
  source: string,
  s: MagicString,
  processor: Processor
) {
  function matchReplacePass(content: string, startOffset: number = 0) {
    const matches = matchRecursive(
      content,
      type.start.pattern,
      type.end.pattern,
      'gmi',
      {
        valueNames: ['between', 'left', 'match', 'right'],
      }
    )

    let left: MatchGroupLeft | null = null
    let match: MatchGroupMatch | null = null
    let right: MatchGroupRight | null = null
    matches.forEach(function ({ name, value, start, end }) {
      start = start + startOffset
      end = end + startOffset
      switch (name) {
        case 'between':
          break
        case 'left':
          left = {
            start,
            end,
            value: type.start.regex.exec(value)!,
          }
          break
        case 'match':
          match = {
            start,
            end,
            value,
          }
          break
        case 'right':
          right = {
            start,
            end,
            value: type.end.regex.exec(value)!,
          }
          if (left && match && right) {
            processor(s, left, right, match, matchReplacePass)
            left = null
            match = null
            right = null
          }
          break
      }
    })
  }

  matchReplacePass(source)
}

function getTestTemplate(test: string) {
  test = test || 'true'
  test = test.trim()

  // force single equals replacement
  // 不替换，会影响 >= 等判断
  // test = test.replace(/([^=!])=([^=])/g, '$1==$2');
  test = test.replace(/-/g, '_')
  return new Function(
    'context',
    'with (context||{}){ return ( ' + test + ' ); }'
  )
}

function testPasses(test: string, context: ProcessContext) {
  var testFn = getTestTemplate(test)
  try {
    return testFn(context)
  } catch (e) {}
  return false
}
