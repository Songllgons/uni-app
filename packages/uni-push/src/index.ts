import GtPush from '../lib/gtpush-min'
import { initPushNotification } from './route'
import { initBroadcastChannel, postPushMessage } from './broadcastChannel'

// if (process.env.UNI_PUSH_DEBUG) {
//   GtPush.setDebugMode(true)
// }

// @ts-expect-error
uni.invokePushCallback({
  type: 'enabled',
})

const appid = process.env.UNI_APP_ID!
if (!appid) {
  Promise.resolve().then(() => {
    // @ts-expect-error
    uni.invokePushCallback({
      type: 'clientId',
      cid: '',
      errMsg: 'manifest.json->appid is required',
    })
  })
} else {
  // #ifdef APP
  initPushNotification()
  // #endif
  // #ifdef H5
  initBroadcastChannel(GtPush)
  // #endif
  // #ifdef MP || APP
  if (typeof uni.onAppShow === 'function') {
    uni.onAppShow(() => {
      GtPush.enableSocket(true)
    })
  }
  // #endif
  GtPush.init({
    appid,
    onError: (res) => {
      console.error(res.error)
      const data = {
        type: 'clientId',
        cid: '',
        errMsg: res.error,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onClientId: (res) => {
      const data = {
        type: 'clientId',
        cid: res.cid,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onlineState: (res) => {
      const data = {
        type: 'lineState',
        online: res.online,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
    onPushMsg: (res) => {
      const data = {
        type: 'pushMsg',
        message: res.message,
      }
      // @ts-expect-error
      uni.invokePushCallback(data)
      // #ifdef H5
      postPushMessage(data)
      // #endif
    },
  })
  // 仅在 jssdk 中监听
  // #ifdef APP
  uni.onPushMessage((res) => {
    if (
      res.type === 'receive' &&
      res.data &&
      (res.data as any).force_notification
    ) {
      // 创建通知栏
      uni.createPushMessage(res.data)
      // 阻止其他监听器继续监听
      ;(res as any).stopped = true
    }
  })
  // #endif
}
