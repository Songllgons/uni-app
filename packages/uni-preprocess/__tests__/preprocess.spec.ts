// import { writeFileSync } from 'node:fs'
// import { resolve } from 'node:path'
import { type SourceMapOptions, preprocess } from '../src/index'

describe('preprocess', () => {
  const sourceMap: SourceMapOptions = {
    includeContent: true,
    hires: true,
    file: 'test',
    source: 'test',
  }

  test('html and js', () => {
    const res = preprocess(
      `<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
  <!-- #ifdef B1 -->
<view>b1</view>
    <!-- #ifdef B11 -->
<view>b11</view>
    <!-- #endif -->
  <!-- #endif -->
  <!-- #ifdef B2 -->
<view>b2</view>
  <!-- #endif -->
<view>b3</view>
<!-- #endif -->
<view>c</view>
<template>
<script>
a
// #ifdef B
b
    // #ifdef B1
b1
        // #ifdef B11
b11
        // #endif
    // #endif
    // #ifdef B2
b2
    // #endif
b3
// #endif
c
</script>
<style>
a
/* #ifdef B */
b
    /* #ifdef B1 */
b1
        /* #ifdef B11 */
b11
        /* #endif */
    /* #endif */
    /* #ifdef B2 */
b2
    /* #endif */
b3
/* #endif */
c
</style>
`,
      {
        context: { B: true, B1: true, B11: true, B2: false },
        sourceMap: sourceMap,
      }
    )
    // 内联sourceMap
    //     writeFileSync(
    //       resolve(__dirname, './test.js'),
    //       `${res.code}
    // //# sourceMappingURL=data:application/json;charset=utf-8;base64,${Buffer.from(
    //         JSON.stringify(res.map)
    //       ).toString('base64')}`
    //     )
    expect(res).toMatchSnapshot()
  })

  describe('html', () => {
    test(`#ifdef match`, () => {
      const res = preprocess(
        `<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
<!-- #endif -->
<view>c</view>
<template>
  `,
        { type: 'html', context: { B: true }, sourceMap: sourceMap }
      )
      expect(res).toMatchSnapshot()
    })
    test(`#ifdef not match`, () => {
      const res = preprocess(
        `<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
<!-- #endif -->
<view>c</view>
<template>
`,
        { type: 'html', context: {}, sourceMap: sourceMap }
      )

      expect(res).toMatchSnapshot()
    })
  })

  describe('js', () => {
    test(`#ifdef match`, () => {
      const res = preprocess(
        `a
// #ifdef B
b
// #endif
c
`,
        { type: 'js', context: { B: true }, sourceMap: sourceMap }
      )
      expect(res).toMatchSnapshot()
    })

    test(`#ifdef not match`, () => {
      const res = preprocess(
        `a
// #ifdef B
b
// #endif
c
`,
        { type: 'js', context: {}, sourceMap: sourceMap }
      )
      expect(res).toMatchSnapshot()
    })

    test(`#ifndef match`, () => {
      const res = preprocess(
        `a
// #ifndef B
b
// #endif
c
`,
        { type: 'js', context: {}, sourceMap: sourceMap }
      )

      expect(res).toMatchSnapshot()
    })

    test(`#ifndef not match`, () => {
      const res = preprocess(
        `a
// #ifndef B
b
// #endif
c
`,
        { type: 'js', context: { B: true }, sourceMap: sourceMap }
      )
      expect(res).toMatchSnapshot()
    })

    test(`#ifdef nesting match`, () => {
      const res = preprocess(
        `a
// #ifdef B
b
    // #ifdef B1
b1
        // #ifdef B11
b11
        // #endif
    // #endif
    // #ifdef B2
b2
    // #endif
b3
// #endif
c
`,
        {
          type: 'js',
          context: { B: true, B1: true, B11: true, B2: false },
          sourceMap: sourceMap,
        }
      )
      expect(res).toMatchSnapshot()
    })

    test(`Unmatched use of #ifdef/#endif.`, () => {
      expect(() => {
        preprocess(
          `a
  // #ifdef B
  b
  // #endif
  // #endif
  c
  `,
          { type: 'js', context: { B: true }, sourceMap: sourceMap }
        )
      }).toThrowErrorMatchingSnapshot()
    })
  })
})
