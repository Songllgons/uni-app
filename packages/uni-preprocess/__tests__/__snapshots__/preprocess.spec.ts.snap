// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preprocess html #ifdef match 1`] = `
{
  "code": "<template>
<view>a</view>
<view>b</view>
<view>c</view>
<template>
  ",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACV,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACV,CAAC",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
<!-- #endif -->
<view>c</view>
<template>
  ",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess html #ifdef not match 1`] = `
{
  "code": "<template>
<view>a</view>
<view>c</view>
<template>
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACV,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAId,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
<!-- #endif -->
<view>c</view>
<template>
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess html and js 1`] = `
{
  "code": "<template>
<view>a</view>
<view>b</view>
<view>b1</view>
<view>b11</view>
<view>b3</view>
<view>c</view>
<template>
<script>
a
b
b1
b11
b3
c
</script>
<style>
a
b
b1
b11
b3
c
</style>
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACV,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEf,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAMhB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAEf,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACV,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACR,CAAC;AAED,CAAC;AAED,CAAC,CAAC;AAEF,CAAC,CAAC,CAAC;AAMH,CAAC,CAAC;AAEF,CAAC;AACD,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACT,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACP,CAAC;AAED,CAAC;AAED,CAAC,CAAC;AAEF,CAAC,CAAC,CAAC;AAMH,CAAC,CAAC;AAEF,CAAC;AACD,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "<template>
<view>a</view>
<!-- #ifdef B -->
<view>b</view>
  <!-- #ifdef B1 -->
<view>b1</view>
    <!-- #ifdef B11 -->
<view>b11</view>
    <!-- #endif -->
  <!-- #endif -->
  <!-- #ifdef B2 -->
<view>b2</view>
  <!-- #endif -->
<view>b3</view>
<!-- #endif -->
<view>c</view>
<template>
<script>
a
// #ifdef B
b
    // #ifdef B1
b1
        // #ifdef B11
b11
        // #endif
    // #endif
    // #ifdef B2
b2
    // #endif
b3
// #endif
c
</script>
<style>
a
/* #ifdef B */
b
    /* #ifdef B1 */
b1
        /* #ifdef B11 */
b11
        /* #endif */
    /* #endif */
    /* #ifdef B2 */
b2
    /* #endif */
b3
/* #endif */
c
</style>
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js #ifdef match 1`] = `
{
  "code": "a
b
c
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC;AAED,CAAC;AAED,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "a
// #ifdef B
b
// #endif
c
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js #ifdef nesting match 1`] = `
{
  "code": "a
b
b1
b11
b3
c
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC;AAED,CAAC;AAED,CAAC,CAAC;AAEF,CAAC,CAAC,CAAC;AAMH,CAAC,CAAC;AAEF,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "a
// #ifdef B
b
    // #ifdef B1
b1
        // #ifdef B11
b11
        // #endif
    // #endif
    // #ifdef B2
b2
    // #endif
b3
// #endif
c
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js #ifdef not match 1`] = `
{
  "code": "a
c
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC;AAID,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "a
// #ifdef B
b
// #endif
c
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js #ifndef match 1`] = `
{
  "code": "a
b
c
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC;AAED,CAAC;AAED,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "a
// #ifndef B
b
// #endif
c
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js #ifndef not match 1`] = `
{
  "code": "a
c
",
  "isInPreprocessor": [Function],
  "map": SourceMap {
    "file": "test",
    "mappings": "AAAA,CAAC;AAID,CAAC;",
    "names": [],
    "sources": [
      "test",
    ],
    "sourcesContent": [
      "a
// #ifndef B
b
// #endif
c
",
    ],
    "version": 3,
  },
}
`;

exports[`preprocess js Unmatched use of #ifdef/#endif. 1`] = `"Unbalanced right delimiter found in string at position 32"`;
